Lab1

KAIFOLOV opened the pull request:
Добавлена многофазная сортировка

KAIFOLOV committed:
Initial commit

KAIFOLOV committed:
Добавил gitignore

KAIFOLOV committed:
убрал билды и пофиксил гит игнор

KAIFOLOV committed:
Cmake тоже убрал

Commit:
добавил Cmake

Commit:
реализовал сортировку слиянием

Commit:
Merge branch 'master' into develop

tvz-041 commented on main.cpp:
```diff
@@ -0,0 +1,117 @@
+        return false;
+    }
+
+    int previousNumber;
+    bool firstNumber = true;
+    int currentNumber;
+
+    while (inFile >> currentNumber) {
+        if (firstNumber) {
+            previousNumber = currentNumber;
+            firstNumber = false;
+            continue;
+        }
+
+        if (currentNumber < previousNumber) {
+            return false;
+        }
+
+        previousNumber = currentNumber;
+    }
+
+    return true;
+}
```
Здесь можно чуть упростить код, если до цикла считать данные в `previousNumber`. Если файл пустой или содержит одно число, функция не зайдёт в цикл и вернёт `true`.
```suggestion
    int previousNumber;
    int currentNumber;
    inFile >> previousNumber;

    while (inFile >> currentNumber) {
        if (currentNumber < previousNumber) {
            return false;
        }

        previousNumber = currentNumber;
    }

    return true;
}
```

KAIFOLOV replied:
Согласен

tvz-041 commented on multiphasesorter.cpp:
```diff
@@ -0,0 +1,130 @@
+    createAuxiliaryFiles(tempFiles, numFiles);
+    splitData(inputFile, tempFiles);
+
+    // Отладочный вывод: содержимое временных файлов после разбиения
+    // std::cout << "Temporary files after splitting:" << std::endl;
+    // for (const auto &tempFile : tempFiles) {
+    //     std::ifstream checkFile(tempFile);
+    //     int val;
+    //     std::cout << "Contents of " << tempFile << ": ";
+    //     while (checkFile >> val) {
+    //         std::cout << val << " ";
+    //     }
+    //     std::cout << std::endl;
+    // }
```
Чтобы не комметировать отладочный вывод, можно завести поле `m_debug` в классе и завернуть подобные конструкции в условие вида `if (m_debug)`.

KAIFOLOV replied:
Согласен, но по хорошему это надо тогда от вида сборки ставить какой-нибудь флаг в cmake, а тут не хотелось привязываться к системе сборки

tvz-041 replied:
Почему? CMake весьма распространён, поддерживается чуть ли не всеми IDE и кроссплатформенный.
Впрочем, как вариант можно использовать флаг `NDEBUG`, на котором завязана работа `assert': https://en.cppreference.com/w/c/error/assert

tvz-041 commented on multiphasesorter.cpp:
```diff
@@ -0,0 +1,130 @@
+
+    // std::cout << "Buffer size set to: " << maxBufferSize << std::endl;
+
+    while (inFile >> number) {
+        buffer.push_back(number);
+
+        if (buffer.size() >= maxBufferSize) {
+            sortAndWriteToFile(buffer, tempFiles[fileIndex]);
+            fileIndex = (fileIndex + 1) % tempFiles.size();
+        }
+    }
+
+    if (!buffer.empty()) {
+        sortAndWriteToFile(buffer, tempFiles[fileIndex]);
+    }
```
Почему здесь происходит сортировка при разбиении?

KAIFOLOV replied:
Чтобы на этапе разбиения иметь уже отсортированные вспомогательные файлы

tvz-041 replied:
Я не о том. Не помню, чтобы в алгоритме многофазной сортировки вообще был такой шаг. Зато были дргуие, которых здесь нет.

KAIFOLOV replied:
Да, не на тот алгоритм я попал при реализации)
Постараюсь переделать в ближайшее время

tvz-041 replied:
У вас же доступ к курсу moodle есть? Там выложены презентации со всеми алгоритмами. Я в своё время постарался сделать их самодостаточными, но если будут вопросы - обращайтесь.

tvz-041 reviewed (commented):
В целом что-то не похоже это на многофазную сортировку. Не видно ни идеального разбиения, ни фиктивных отрезков.

Commit:
Добавил разбиение на серии по числам фибоначи
Добавил фейковые серии

tvz-041 commented on multiphasesorter.cpp:
```diff
@@ -0,0 +1,180 @@
+    cleanupTempFiles(tempFiles);
+}
+
+std::vector<int> MultiPhaseSorter::generateFibonacciDistribution(const int numFiles)
+{
+    std::vector<int> fibSeries = { 1, 1 };
+    while (fibSeries.size() < numFiles) {
+        fibSeries.push_back(fibSeries[fibSeries.size() - 1] + fibSeries[fibSeries.size() - 2]);
+    }
+    return fibSeries;
+}
```
Зачем генерировать распределение?
В лекции был предложен алгоритм, который позволяет строить следующее распределение отрезков по текущему.

tvz-041 commented on multiphasesorter.cpp:
```diff
@@ -0,0 +1,180 @@
+        }
+
+        if (!buffer.empty()) {
+            std::sort(buffer.begin(), buffer.end());
+            writeToFile(buffer, tempFiles[fileIndex]);
```
Я всё ещё не понимаю, почему здесь происходит сортировка буфера произвольного размера, если во всех файловх сортировках не требуется держать в памяти более 2-х значений единовременно при чтении из исходного файла.

tvz-041 commented on multiphasesorter.cpp:
```diff
@@ -0,0 +1,180 @@
+            writeToFile(buffer, tempFiles[fileIndex]);
+        }
+
+        while (buffer.size() < currentSeriesSize) {
+            addFakeSegment(tempFiles[fileIndex]);
+            ++currentSeriesSize;
+        }
```
В этом случае не будет горизонтального распределения фиктивных отрезком (см. лекции).

tvz-041 commented on multiphasesorter.cpp:
```diff
@@ -0,0 +1,180 @@
+        int currentSeriesSize = fibSeries[fileIndex];
+        buffer.clear();
+
+        while (buffer.size() < currentSeriesSize && inFile >> number) {
```
Длина подпоследовательности вообще никакого отношения к числам Фибоначчи не имеет.

tvz-041 commented on multiphasesorter.cpp:
```diff
@@ -0,0 +1,180 @@
+    for (int i = 0; i < n; ++i) {
+        fileStreams.emplace_back(tempFiles[i]);
+        if (!fileStreams.back().is_open()) {
+            std::cerr << "Error: Could not open temp file " << tempFiles[i] << std::endl;
+            return;
```
Имеет смысл сделать функцию `bool` и возвращать здесь `false`.

Commit:
полный рефакторинг

Commit:
правильное разбиение

Commit:
теперь точно правильное разбиение

Commit:
Отказался от идеи статичной сортировки
Добавил массивы для отслеживания фиктивных отрезков и разбиения
с парочкой костылей, но готовая многофазная сортировка

tvz-041 commented on multiphasesorter.cpp:
```diff
@@ -0,0 +1,295 @@
+
+void MultiPhaseSorter::cleanupTempFiles()
+{
+    for (const auto &fileName : _tempFiles) {
+        std::remove(fileName.c_str());
+    }
```
Вероятно, где-то здесь имеет смысл почистить `_tempFiles` (весь класс, как я понял, спроектирован под повторный вызов сортировки).

KAIFOLOV replied:
пофикшено

tvz-041 commented on multiphasesorter.cpp:
```diff
@@ -0,0 +1,295 @@
+        if (!tempFileStreams[i].is_open()) {
+            std::cerr << "Error: Could not open temp file " << _tempFiles[i] << " for writing.\n";
+            return;
+        }
```
Здесь тоже напрашивается `return false` с тем чтобы основная функция не пыталась делать слияние, если не удалось провести разбиение.

KAIFOLOV replied:
пофикшено

tvz-041 commented on multiphasesorter.cpp:
```diff
@@ -0,0 +1,295 @@
+
+    int i = 0;
+    int currentNumber, lastNumber;
+    inFile >> currentNumber;
```
Исходный файл может быть пустым.

KAIFOLOV replied:
пофикшено

tvz-041 replied:
Можно было просто 69 строку в условие завернуть вместо seekg, ну да ладно.
После прочтения ответа тему можете resolv'ить.

tvz-041 commented on multiphasesorter.cpp:
```diff
@@ -0,0 +1,295 @@
+    std::vector<std::fstream *> tempFiles(_numFiles);
+
+    for (int i = 0; i < _numFiles - 1; ++i) {
+        tempFiles[i] = new std::fstream(_tempFiles[i], std::ios::in);
```
По-моему, эта память в итоге нигде не освобождается.

KAIFOLOV replied:
пофикшено

tvz-041 replied:
В лекции этого не было, но у потоков есть метод `swap`, который позмоляет обменивать содержимое местами:
```c++
    std::vector<std::fstream> files(10);
    for (int i = files.size() - 1; i > 0; --i)
    {
        files[i].swap(files[i - 1]);
    }
```
Так что если в будущем будет похожая задача, можно обойтись без динамической памяти.
После прочтения ответа тему можете resolv'ить.

tvz-041 commented on multiphasesorter.cpp:
```diff
@@ -0,0 +1,295 @@
+                if (_ip[i] <= 0)
+                    continue;
+
+                // КОСТЫЛЬ: В СИТУАЦИИ КОГДА РАБОТАЕМ С МНИМЫМИ ФРАГМЕНТАМИ, РАЗДЕЛИТЕЛЬ ПИШЕТСЯ В
+                // ФАЙЛ, ПРИ ЧТЕНИИ ЕГО ОТБРАСЫВАЕМ
+                int tmp;
+                do {
+                    *files[i] >> tmp;
+
+                    if (tmp == SEGMENT_DELIMITER) {
+                        _ip[i]--;
+                    }
+
+                } while (tmp == SEGMENT_DELIMITER && _ip[i] > 0);
```
Странно, я что-то нигде не вижу, чтобы вы записывали пустые отрезки в файл.

KAIFOLOV replied:
Я не записываю фиктивные (пустые) отрезки в файл, а просто помечаю в массиве _ms что в таком-то файле столько-то пустых отрезков

tvz-041 replied:
Ну так в комментарии написано, что разделитель фиктивных отрезков в файл пишется.

KAIFOLOV replied:
Этот костыль имеет прямое отношение к треду ниже, попробую избавиться

KAIFOLOV replied:
Собственно получается что когда я пишу фиктивный отрезок в сливаемый файл, я пишу туда еще просто разделитель, и получается ситуация когда подряд идут несколько разделителей, так вот этот костыль решает эту проблему

tvz-041 replied:
А, на 256 строке? В целом, если вы выполните предыдущую рекомендацию, то в том же условии можно будет просто поставить `continue`. Мы ведь когда увеличили `ms`, после этого вроде ничего делать не нужно, переходим к следующей итерации.

KAIFOLOV replied:
тоже решил

tvz-041 commented on multiphasesorter.cpp:
```diff
@@ -0,0 +1,295 @@
+        bool hasRealData = false;
+        // КОСТЫЛЬ: ЕСЛИ ВО ВСЕХ ФАЙЛАХ ЕСТЬ МНИМЫЙ ОТРЕЗОК, ТО В СЛИВАЕМЫЙ ФАЙЛ ТОЖЕ НАДО ПРИБАВИТЬ
+        // МНИМЫЙ
+        std::vector<bool> costul(_numFiles - 1, false);
```
А зачем нам запоминать, в каких файлах были пустые отрезки? Нам же достаточно знать, что они во всех/не во всех файлах.

KAIFOLOV replied:
тут как раз ситуация, когда например при слиянии в каждом файле есть фиктивный отрезок, то в сливаемый файл я тоже плюсую фиктивный отрезок

tvz-041 replied:
Это понятно. Зачем вам вектор, если достаточно одной переменной `bool`?

KAIFOLOV replied:
в принципе справедливо, под конец я пытался изобрести эту конструкцию, как раз кусок в виде флага "hasRealData" от нее остался, но чета не получалось, надо свежим взглядом попробовать еще

KAIFOLOV replied:
Вспомнил зачем это нужно) мне нужно только в конце, после просмотра всех файлов плюсануть фиктивный отрезок в сливаемый, а с одним флагом bool у меня получается что он после проверки каждого файла плюсует, в общем у меня нет идей как это победить

tvz-041 replied:
А зачем условие на 232 строке вообще держать в этом цикле?
Этот код всё равно в конце отрабатывает, когда все итерации по факту пройдены.

KAIFOLOV replied:
да, все окей, я разобрался и убрал костыль

tvz-041 reviewed (commented):
Просьба на будущее не resolv'ить треды, а просто там отписываться. Мне просто всё равно их потом просматривать нужно, чтобы вспомнить что было :)
Если вопросов к треду не будет, сам уже помечу решённым.

tvz-041 commented:
А, ещё моментик. У нас эти лабораторные можно поделить на 4 группы:
- файловые сортировки
- деревья (2-3 лабы)
- кодирование
- хеширование
И в целом, их можно независимо делать в отдельных ветках.
Поэтому как вариант, если время есть, в ожидании ревью можно на другие лабораторные переключаться.

Commit:
Добавлена проверка на пустой файл
добавлена очистка вектора имен вспомогательных файлов
добавлена проверка на успешное разбиение

KAIFOLOV commented:
Такой вопрос, в целом у меня не так много времени, по причине работы, на оценку 3 надо прям все лабы сделать? А то что-то процесс затягивается)

tvz-041 commented on multiphasesorter.h:
```diff
@@ -0,0 +1,41 @@
+#include <limits>
+#include <vector>
+
+constexpr int SEGMENT_DELIMITER = std::numeric_limits<int>::min();
```
Лучше инкапсулировать в класс как `static` поле, на пределами класса нам эта константа не нужна.

KAIFOLOV replied:
пофикшено

tvz-041 commented:
> Такой вопрос, в целом у меня не так много времени, по причине работы, на оценку 3 надо прям все лабы сделать? А то что-то процесс затягивается)

Нет, не все. В задании обычно первой строчкой указывается, если выполенеие лабораторной для какой-то оценки не требуется. По-моему, во втором семестре это лабораторная №4.
Также, в зависимости от ценки, идёт разный объём по методам, они как правило помечены цветом.

Commit:
Разделитель больше не пишется при слиянии фиктивных отрезков

Commit:
занес разделитель в поле класса

tvz-041 commented:
После аппрува PR, его можно сливать. Мало ли, вдруг не заметили)
